File: android/app/src/main/java/com/styloai/app/ui/screens/chat/ChatViewModel.kt
package com.styloai.app.ui.screens.chat

import androidx.lifecycle.viewModelScope
import com.styloai.app.data.model.*
import com.styloai.app.data.repository.ChatRepository
import com.styloai.app.ui.base.BaseViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import java.util.UUID
import javax.inject.Inject

data class ChatUiState(
    val messages: List<DisplayMessage> = listOf(
        DisplayMessage(
            id = "welcome",
            role = MessageRole.ASSISTANT,
            content = "Hi! I'm your AI style assistant. Ask me anything about fashion, outfit ideas, or get personalized recommendations based on your wardrobe!",
            quickActions = listOf(
                QuickAction("Outfit for today", "suggest_outfit_today"),
                QuickAction("Style tips", "style_tips"),
                QuickAction("Color matching", "color_help")
            )
        )
    ),
    val inputText: String = "",
    val conversationId: String? = null,
    val conversations: List<Conversation> = emptyList()
)

@HiltViewModel
class ChatViewModel @Inject constructor(
    private val repository: ChatRepository
) : BaseViewModel<ChatUiState>(ChatUiState()) {

    fun updateInputText(text: String) {
        updateState { it.copy(inputText = text) }
    }

    fun sendMessage() {
        val text = uiState.value.inputText.trim()
        if (text.isEmpty()) return

        val userMessage = DisplayMessage(
            id = UUID.randomUUID().toString(),
            role = MessageRole.USER,
            content = text
        )

        updateState { 
            it.copy(
                messages = it.messages + userMessage,
                inputText = ""
            )
        }

        viewModelScope.launch {
            setLoading(true)
            val result = repository.sendMessage(
                message = text,
                conversationId = uiState.value.conversationId
            )

            result.fold(
                onSuccess = { response ->
                    val assistantMessage = DisplayMessage(
                        id = response.message.id,
                        role = MessageRole.ASSISTANT,
                        content = response.message.content,
                        quickActions = response.message.quickActions
                    )

                    updateState { 
                        it.copy(
                            messages = it.messages + assistantMessage,
                            conversationId = response.conversationId
                        )
                    }
                    setLoading(false)
                },
                onFailure = { error ->
                    val errorMessage = DisplayMessage(
                        id = UUID.randomUUID().toString(),
                        role = MessageRole.ASSISTANT,
                        content = "I'm sorry, I encountered an error. Please try again."
                    )

                    updateState { it.copy(messages = it.messages + errorMessage) }
                    setLoading(false)
                    showError(error.message ?: "Failed to send message")
                }
            )
        }
    }

    fun handleQuickAction(action: String) {
        val message = when (action) {
            "suggest_outfit_today" -> "Suggest an outfit for today based on the weather"
            "style_tips" -> "Give me some style tips for my body type"
            "color_help" -> "Help me understand color matching in outfits"
            else -> action
        }

        updateInputText(message)
        sendMessage()
    }

    fun startNewConversation() {
        updateState { ChatUiState() }
    }

    fun loadConversations() {
        viewModelScope.launch {
            repository.getConversations().onSuccess { conversations ->
                updateState { it.copy(conversations = conversations) }
            }.onFailure { error ->
                showError(error.message ?: "Failed to load conversations")
            }
        }
    }

    fun loadConversation(conversationId: String) {
        viewModelScope.launch {
            setLoading(true)

            repository.getMessages(conversationId).fold(
                onSuccess = { messages ->
                    val displayMessages = messages.map { msg ->
                        DisplayMessage(
                            id = msg.id,
                            role = msg.role,
                            content = msg.content,
                            timestamp = msg.timestamp,
                            quickActions = msg.quickActions
                        )
                    }

                    updateState { 
                        it.copy(
                            messages = displayMessages,
                            conversationId = conversationId
                        )
                    }
                    setLoading(false)
                },
                onFailure = { error ->
                    setLoading(false)
                    showError(error.message ?: "Failed to load conversation")
                }
            )
        }
    }
}

File: ios/StyloAI/ViewModels/ChatViewModel.swift
import SwiftUI

struct DisplayMessage: Identifiable {
    let id = UUID()
    let role: MessageRole
    let content: String
    let quickActions: [QuickAction]?
    let timestamp: Date

    init(role: MessageRole, content: String, quickActions: [QuickAction]? = nil) {
        self.role = role
        self.content = content
        self.quickActions = quickActions
        self.timestamp = Date()
    }

    init(from chatMessage: ChatMessage) {
        self.role = chatMessage.role
        self.content = chatMessage.content
        self.quickActions = nil
        self.timestamp = chatMessage.createdAt
    }
}

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [DisplayMessage] = []
    @Published var inputText = ""
    @Published var isLoading = false
    @Published var conversationId: String?
    @Published var conversations: [Conversation] = []

    init() {
        addWelcomeMessage()
    }

    private func addWelcomeMessage() {
        let welcome = DisplayMessage(
            role: .assistant,
            content: "Hi! I'm your personal style assistant. I can help you with outfit recommendations, style advice, and making the most of your wardrobe. What would you like help with today?",
            quickActions: [
                QuickAction(label: "Today's Weather Picks", action: "weather_outfit"),
                QuickAction(label: "Outfit for Event", action: "occasion_outfit"),
                QuickAction(label: "Style Tips", action: "style_tips"),
                QuickAction(label: "Wardrobe Stats", action: "wardrobe_stats")
            ]
        )
        messages.append(welcome)
    }

    func sendMessage() async {
        guard !inputText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        let userMessage = inputText
        inputText = ""

        // Add user message
        messages.append(DisplayMessage(role: .user, content: userMessage))

        isLoading = true

        do {
            let response = try await APIService.shared.sendChatMessage(
                message: userMessage,
                conversationId: conversationId
            )

            conversationId = response.conversationId

            // Add assistant response
            messages.append(DisplayMessage(
                role: .assistant,
                content: response.message,
                quickActions: response.quickActions
            ))
        } catch {
            messages.append(DisplayMessage(
                role: .assistant,
                content: "Sorry, I couldn't process your request. Please try again."
            ))
        }

        isLoading = false
    }

    func handleQuickAction(_ action: String) async {
        let actionMessages: [String: String] = [
            "weather_outfit": "What should I wear based on today's weather?",
            "occasion_outfit": "I need an outfit for an event",
            "style_tips": "Give me some style tips",
            "wardrobe_stats": "Tell me about my wardrobe",
            "add_item": "I want to add a new item to my wardrobe",
            "show_outfits": "Show me some outfit ideas",
            "analyze_style": "Analyze my style preferences"
        ]

        if let message = actionMessages[action] {
            inputText = message
            await sendMessage()
        }
    }

    func loadConversations() async {
        do {
            conversations = try await APIService.shared.getConversations()
        } catch {
            print("Failed to load conversations: \(error)")
        }
    }

    func loadConversation(_ id: String) async {
        conversationId = id
        messages = []

        do {
            let chatMessages = try await APIService.shared.getConversation(id: id)
            messages = chatMessages.map { DisplayMessage(from: $0) }
        } catch {
            addWelcomeMessage()
        }
    }

    func startNewConversation() {
        conversationId = nil
        messages = []
        addWelcomeMessage()
    }
}
